File: apps/android/app/src/main/java/dev/flextranslate/foundation/GeminiFlashTranslationProvider.kt
package dev.flextranslate.foundation

/**
 * Реальный облачный MT-уровень WS5 за кандидатом `gemini-flash-cloud` в пикере.
 *
 * Два режима credential (см. [GeminiCredentialMode]):
 *
 * **BACKEND_MEDIATION** (изначальный путь): перевод идёт через НАШ mediation-бэкенд
 * ([CloudMediationClient]), который подставляет credential Gemini на сервере и возвращает только
 * текст. Приложение ключ Gemini вообще не держит. Гейтится через [CloudCallGate] (согласие +
 * раскрытие + online + ephemeral-токен + настроенный бэкенд).
 *
 * **OWN_KEY** (BYOK): собственный API-ключ Gemini юзера (хранится зашифрованным в [GeminiKeyStore])
 * достаётся в момент вызова и уходит в `x-goog-api-key` напрямую на REST-эндпоинт Gemini через
 * [GeminiDirectClient]. Тоже гейтится (согласие + раскрытие + online + есть ключ). Гео-ограничение
 * (HTTP 400 "User location is not supported") показываем честно — не делаем вид, что вызов прошёл
 * там, где Gemini заблокирован по региону.
 *
 * Контракт честности (§1.2 / §3): заблокированный гейтом, офлайновый, отклонённый или упавший вызов
 * возвращает `TranslationResult(text = null, unsupportedReason = <причина на языке продукта>)`.
 * Никогда не выдумывает вывод и не сваливается молча на on-device модель.
 *
 * Id модели берётся из конфига ([GeminiFlashConfig.modelId], по умолчанию `gemini-3.5-flash`).
 */
class GeminiFlashTranslationProvider(
    private val config: GeminiFlashConfig,
    private val gate: CloudCallGate,
    private val backend: CloudMediationClient,
    private val directClient: GeminiDirectClient = GeminiDirectClient(config),
    private val keyStore: GeminiKeyStore? = null,
    override val providerId: String = PROVIDER_ID,
    private val clock: () -> Long = System::currentTimeMillis,
) : TranslationProvider {

    override fun translate(text: String, languagePair: String, deviceTier: String): TranslationResult {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return TranslationResult(text = null, unsupportedReason = null)

        // Только заявленные демо-направления с пивотом через RU; всё за рамками отбрасываем до того,
        // как тратить облачный вызов.
        MtDirection.parse(languagePair)
            ?: return TranslationResult(
                text = null,
                unsupportedReason = "пара $languagePair не поддерживается облачной моделью",
            )

        // 1. Жёсткий гейт. Нет согласия / раскрытия / офлайн → вызова нет (оба режима).
        //    Плюс: нет бэкенда → блок (mediation); нет ключа → блок (own-key).
        val decision = gate.evaluate(providerId, clock())
        if (decision is CloudCallGate.Decision.Blocked) {
            return TranslationResult(text = null, unsupportedReason = decision.userReason)
        }
        val allowed = decision as CloudCallGate.Decision.Allowed

        // 2. Выбираем транспорт по режиму credential.
        return when (config.credentialMode) {
            GeminiCredentialMode.OWN_KEY -> translateDirect(trimmed, languagePair)
            GeminiCredentialMode.BACKEND_MEDIATION -> translateMediated(trimmed, languagePair, allowed)
        }
    }

    /** Путь OWN_KEY: достаём ключ из защищённого хранилища в момент вызова, бьём в Gemini напрямую. */
    private fun translateDirect(text: String, languagePair: String): TranslationResult {
        // Ключ берём только на время вызова — между вызовами не держим.
        val apiKey = keyStore?.loadKey()
        if (apiKey.isNullOrBlank()) {
            return TranslationResult(text = null, unsupportedReason = CloudCallGate.REASON_NO_OWN_KEY)
        }
        return when (val result = directClient.translate(text, languagePair, apiKey)) {
            is GeminiDirectClient.DirectResult.Ok ->
                TranslationResult(text = result.text, unsupportedReason = null)
            GeminiDirectClient.DirectResult.GeoBlocked ->
                TranslationResult(text = null, unsupportedReason = REASON_GEO_BLOCKED)
            GeminiDirectClient.DirectResult.KeyRejected ->
                TranslationResult(text = null, unsupportedReason = REASON_KEY_REJECTED)
            is GeminiDirectClient.DirectResult.Failed ->
                TranslationResult(text = null, unsupportedReason = REASON_DIRECT_FAILED)
        }
    }

    /** Путь BACKEND_MEDIATION: собираем запрос-намерение и отдаём его mediation-бэкенду. */
    private fun translateMediated(
        text: String,
        languagePair: String,
        allowed: CloudCallGate.Decision.Allowed,
    ): TranslationResult {
        val request = GeminiTranslateRequest(
            providerId = providerId,
            modelId = config.modelId,
            languagePair = languagePair,
            text = text,
            thinkingLevel = config.thinkingLevel,
            stream = config.streaming,
        )
        return when (val result = backend.translate(request, allowed.credential)) {
            is CloudMediationClient.Result.Ok ->
                TranslationResult(text = result.text, unsupportedReason = null)
            is CloudMediationClient.Result.Refused ->
                TranslationResult(text = null, unsupportedReason = result.userReason)
            is CloudMediationClient.Result.Failed ->
                TranslationResult(text = null, unsupportedReason = "Облачный перевод недоступен")
        }
    }

    companion object {
        const val PROVIDER_ID = "gemini-flash-cloud"

        // Честные причины для прямого пути OWN_KEY — показываются на экране Live.
        const val REASON_GEO_BLOCKED =
            "Gemini недоступен в вашем регионе — используйте backend-режим или VPN"
        const val REASON_KEY_REJECTED =
            "Gemini API-ключ отклонён (проверьте ключ в настройках Облака)"
        const val REASON_DIRECT_FAILED =
            "Прямой вызов Gemini не удался"
    }
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/GeminiFlashTranslationProvider.swift
import Foundation

/// The real WS5 cloud MT tier behind the gemini-flash-cloud picker candidate.
///
/// Supports two credential modes (see GeminiCredentialMode):
///
/// backendMediation (original path): translation goes through OUR mediation backend, which
/// injects the Gemini credential server-side and returns text only. The app never holds a
/// Gemini key. Gated by CloudCallGate (consent + disclosure + online + ephemeral token +
/// backend configured).
///
/// ownKey (BYOK): the user's own Gemini API key (stored in iOS Keychain, never UserDefaults)
/// is fetched just-in-time and sent in x-goog-api-key directly to the Gemini REST endpoint
/// via GeminiDirectClient. Still gated (consent + disclosure + online + key present).
/// Geo-restriction (HTTP 400 "User location is not supported") is surfaced honestly.
///
/// Honesty contract: a gated, offline, declined, or failed call returns
/// TranslationResult(text: nil, unsupportedReason: <reason>). Never fabricates output,
/// never silently falls back to an on-device model.
///
/// Model id is config-driven (GeminiFlashConfig.modelId, default gemini-3.5-flash).
///
/// Mirrors Android GeminiFlashTranslationProvider.
final class GeminiFlashTranslationProvider: TranslationProvider, @unchecked Sendable {

    static let providerId = "gemini-flash-cloud"

    // Honest reasons for the ownKey direct path.
    static let reasonGeoBlocked =
        "Gemini is not available in your region — use backend mode or a VPN"
    static let reasonKeyRejected =
        "Gemini API key rejected (check the key in Cloud settings)"
    static let reasonDirectFailed =
        "Direct Gemini call failed"

    let providerId: String = GeminiFlashTranslationProvider.providerId

    private let config: GeminiFlashConfig
    private let gate: CloudCallGate
    private let backend: any CloudMediationClient
    private let directClient: GeminiDirectClient
    private let keyStore: (any GeminiKeyStore)?
    private let clock: @Sendable () -> Int64

    init(
        config: GeminiFlashConfig,
        gate: CloudCallGate,
        backend: any CloudMediationClient,
        directClient: GeminiDirectClient? = nil,
        keyStore: (any GeminiKeyStore)? = nil,
        clock: @escaping @Sendable () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.config = config
        self.gate = gate
        self.backend = backend
        self.directClient = directClient ?? GeminiDirectClient(config: config)
        self.keyStore = keyStore
        self.clock = clock
    }

    func translate(text: String, languagePair: String, deviceTier: String) -> TranslationResult {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return TranslationResult(text: nil, unsupportedReason: nil) }

        // Restrict to advertised RU-pivot demo directions (uses existing MtDirection from M2m100MtProvider).
        guard MtDirection.parse(languagePair) != nil else {
            return TranslationResult(
                text: nil,
                unsupportedReason: "pair \(languagePair) not supported by cloud model"
            )
        }

        // Hard gate: no consent / no disclosure / offline → no call (both modes).
        let decision = gate.evaluate(providerId: Self.providerId, nowEpochMs: clock())
        switch decision {
        case .blocked(let reason):
            return TranslationResult(text: nil, unsupportedReason: reason)
        case .allowed(let credential):
            switch config.credentialMode {
            case .ownKey:
                return translateDirect(trimmed, languagePair: languagePair)
            case .backendMediation:
                return translateMediated(trimmed, languagePair: languagePair, credential: credential)
            }
        }
    }

    /// ownKey path: fetch key from Keychain just-in-time, call Gemini directly.
    private func translateDirect(_ text: String, languagePair: String) -> TranslationResult {
        guard let apiKey = keyStore?.loadKey(), !apiKey.isEmpty else {
            return TranslationResult(text: nil, unsupportedReason: CloudCallGate.reasonNoOwnKey)
        }
        switch directClient.translate(text: text, languagePair: languagePair, apiKey: apiKey) {
        case .ok(let translated):
            return TranslationResult(text: translated, unsupportedReason: nil)
        case .geoBlocked:
            return TranslationResult(text: nil, unsupportedReason: Self.reasonGeoBlocked)
        case .keyRejected:
            return TranslationResult(text: nil, unsupportedReason: Self.reasonKeyRejected)
        case .failed:
            return TranslationResult(text: nil, unsupportedReason: Self.reasonDirectFailed)
        }
    }

    /// backendMediation path: build intent request and delegate to the mediation backend.
    private func translateMediated(
        _ text: String,
        languagePair: String,
        credential: CloudCredential
    ) -> TranslationResult {
        let request = GeminiTranslateRequest(
            providerId: Self.providerId,
            modelId: config.modelId,
            languagePair: languagePair,
            text: text,
            thinkingLevel: config.thinkingLevel,
            stream: config.streaming
        )
        switch backend.translate(request: request, credential: credential) {
        case .ok(let translated, _):
            return TranslationResult(text: translated, unsupportedReason: nil)
        case .refused(let reason):
            return TranslationResult(text: nil, unsupportedReason: reason)
        case .failed:
            return TranslationResult(text: nil, unsupportedReason: "Cloud translation unavailable")
        }
    }
}

